test/pytest/service-desktop/test_user_data_backup_restore.py

```diff
@@ -0,0 +1,268 @@
+            log.debug("Restore done, rebooting")
+    else:
+        log.error("No backup files on phone")
+        raise TestError(Error.OTHER_ERROR)
+
+def populate_user_data(harness):
+    contacts = GetContactsCount().run(harness).count
+    log.debug(f"Total contacts: {contacts}")
+
+    if contacts == 0:
+        resp = AddContact({
+            "altName": "Testowy",
+            "favourite": True,
+            "numbers": ["547623521"],
+            "priName": "Test"}).run(harness)
+
+        log.debug(f"Created contact id: {resp.id}")
+        contacts += 1
+
+        resp = AddContact({
+            "altName": "Testowy2",
+            "favourite": True,
+            "numbers": ["123456789"],
+            "priName": "Test2"}).run(harness)
+
+        log.debug(f"Created contact id: {resp.id}" )
+        contacts += 1
+
+    return contacts
+
+def clear_user_data(harness):
+    for id in AddedContactIds:
+        DeleteContactById(id).run(harness)
+        log.debug(f"Deleted contact id: {id}")
+
+    log.debug("Confirm no user data on device")
+
+    contactsCountAfterClear = GetContactsCount().run(harness).count
+    log.debug(f"Contacts after clear: {contactsCountAfterClear}")
+
+    assert contactsCountAfterClear == 0
+
+def save_user_data(harness):
+    log.debug("Save user data from device for comparison")
+    resp = GetContactsWithOffsetAndLimit(0, totalContacts).run(harness)
+
+    for e in resp.entries:
+        id = e["id"]
+        ReferenceContacts[id] = e
+        AddedContactIds.append(id)
+        log.debug(f"Saved contact id: {id}")
+
+
+def compare_user_data(harness, totalContacts):
+    contactsCountAfterRestore = GetContactsCount().run(harness).count
+
+    log.debug(f"Contacts after restore: {contactsCountAfterRestore}")
+    assert totalContacts == contactsCountAfterRestore
+
+    log.debug("Get user data from device for comparison")
+    resp = GetContactsWithOffsetAndLimit(0, totalContacts).run(harness)
+
+    for e in resp.entries:
+        id = e["id"]
+        log.debug(f"Compare contact id: {id}")
+        assert ReferenceContacts[id] == e
+        assert id in AddedContactIds
+
+
+@pytest.mark.service_desktop_test
+@pytest.mark.rt1051
+@pytest.mark.usefixtures("phone_unlocked")
+@pytest.mark.backup
+def test_backup_and_restore(harness):
+    log.debug("Checking/Creating user data on device")
+    totalContacts = populate_user_data(harness)
+
+    log.debug("Save user data from device for comparison")
+    save_user_data(harness)
+
+    remove_stalled_backups(harness)
+
+    log.debug("Creating backup")
+    run_backup(harness)
+
+    log.debug("Removing user data from device")
+    clear_user_data(harness)
+
+    log.debug("Requesting restore")
+    perform_restore(harness)
+
+    harness.connection.watch_port_reboot(20)
+    harness = get_harness_automatic(20)
+    harness.unlock_phone()
+
+    log.debug("Confirm user data on device match saved data")
+    compare_user_data(harness, totalContacts)
```

Comment:
What about restoring the initial state of the DUT?
We want to leave DUT in the same state as it was before each test (eg. no extra user data should be left on the DUT).

Reply:
Good point, fixing.